architecture-docs/adr-002.md

```diff
@@ -0,0 +1,490 @@
+# ADR 002: Finalize specifications for interaction between Validator Set Changes, Punishments and Rewards
+
+## Changelog
+- 23-01-2020: Initial Draft
+
+## Context
+
+Currently, `ValidatorState` is used by punishments and rewards as a source of data to compute their respective slashing
+ratio and rewards amount. Besides this, `ValidatorState` is also used to manage dynamic validator set (i.e.,
+validators can be added or removed based on their network operations).
+
+### Problems with current design
+
+1. Our current design does not delete validator information from `ValidatorState` when a validator either unbonds their
+   funds or gets removed from validator set because of some punishment. The reason for not deleting this information is
+   because it is still required for computing `SlashRatio` while slashing a punished validator. Because of this,
+   validators cannot rejoin validator set (because their information already exists and we cannot add a duplicate
+   entry).
+1. During rewards calculation, we calculate `total_stake` based of the active validator at the time of rewards
+   distribution. This does not take into account all the validator set changes that took place in current rewards
+   period. For example, when a validator unbonds or deposits funds just before rewards distribution (let us assume that
+   they unbonded their funds at 23rd hour), their stake is not considered in rewards calculation.
+
+Besides this, validator set, rewards pool and punishments are three different components of our system and should
+maintain their own data instead of relying on `ValidatorState` for information. There should be a specified
+interface/behavior for interaction between these components.
```

Comment:
If we keep several copied of the same data, extra care should be taken to keep them in sync.

Reply:
If you go through this document, you’ll see that we’re not keeping separate copies of same data, we’re storing data where it is needed and deleting when it is no longer needed.